### PR TITLE
Support json type

### LINKED ISSUE #270: Support JSON types
- [x] Add type conversion in common scalar types.
- [x] Support generate correct types in codegen
- [x] Support in query 

Preferred pattern : 

```graphql

type starterEntity @entity {
  field6: JsonField
  field7: [JsonField]
  field8: [String]
}


type JsonField @jsonField {
  field1: Int!
  field2: BigInt
  field3: []
}



```

### PR DESCRIPTION
**Pull request by jiqiang90**
Closes #270

### CODE REVIEW

File: packages/cli/src/controller/codegen-controller.ts [hunk 1]
```diff
@@ -4,20 +4,33 @@
...
 const MODEL_ROOT_DIR = 'src/types/models';
+const exportTypes = {
+  models: false,
+  interfaces: false,
+};
 
 // 4. Render entity data in ejs template and write it
-export async function renderTemplate(outputPath: string, templateData: ejs.Data): Promise<void> {
-  const data = await ejs.renderFile(TEMPLATE_PATH, templateData);
+export async function renderTemplate(template_path: string, outputPath: string, templateData: ejs.Data): Promise<void> {
```
Comment (ianhe8x) [inline packages/cli/src/controller/codegen-controller.ts:32]: camel case for variable name

File: packages/cli/src/controller/codegen-controller.ts [hunk 2]
```diff
@@ -26,36 +39,102 @@ export interface processedField {
...
+      await renderTemplate(INTERFACE_TEMPLATE_PATH, path.join(typesDir, `interfaces.ts`), interfaceTemplate);
+      exportTypes.interfaces = true;
+    } catch (e) {
+      throw new Error(`When render json interfaces having problems.`);
+    }
+  }
 }
 
-export function processFields(className: string, fields: GraphQLEntityField[]): processedField[] {
+export function processFields(type: string, className: string, fields: GraphQLEntityField[] | GraphQLJsonFieldType[]) {
```
Comment (ianhe8x) [inline packages/cli/src/controller/codegen-controller.ts:73]: can't find where `type` is used

Comment (ianhe8x) [inline packages/cli/src/controller/codegen-controller.ts:73]: `GraphQLEntityField[] | GraphQLJsonFieldType[]` should be more like 
`(GraphQLEntityField | GraphQLJsonFieldType)[]`

Comment (jiqiang90) [reply packages/cli/src/controller/codegen-controller.ts]: Two types 'entity' and 'jsonField' were pass though type for throw Error after transform field type , so the error message is more accurate. Also currently only 'entity' type need to extract indexed fields.

Comment (jiqiang90) [reply packages/cli/src/controller/codegen-controller.ts]: Fixed

File: packages/cli/src/controller/codegen-controller.ts (hunk 1 above)
Comment (jiqiang90) [reply packages/cli/src/controller/codegen-controller.ts]: Fixed

File: packages/cli/src/controller/codegen-controller.ts [hunk 3]
```diff
@@ -28,55 +41,128 @@ export interface ProcessedField {
...
       name: field.name,
-      type: newType,
       required: !field.nullable,
-      indexed,
-      unique,
-    });
+      isArray: field.isArray,
+    } as ProcessedField;
+    if (type === 'entity' && indexFields) {
+      const [indexed, unique] = indexFields.reduce<[boolean, boolean]>(
```
Comment (ianhe8x) [inline packages/cli/src/controller/codegen-controller.ts:92]: these several lines need to run for type: entity no matter indexFields exists or not.
though you can give indexFields a default value `[]`

File: packages/cli/src/controller/codegen-controller.ts [hunk 4]
```diff
@@ -28,55 +41,128 @@ export interface ProcessedField {
...
+    try {
+      await renderTemplate(TYPES_INDEX_TEMPLATE_PATH, path.join(projectPath, TYPE_ROOT_DIR, `index.ts`), {
+        props: {
+          exportTypes,
+        },
+      });
+    } catch (e) {
+      throw new Error(`When render index in types having problems.`);
+    }
+    console.log(`* Types index generated !`);
```
Comment (ianhe8x) [inline packages/cli/src/controller/codegen-controller.ts:164]: no console.log

File: packages/cli/src/template/interface.ts.ejs [hunk 5]
```diff
@@ -0,0 +1,12 @@
+// Copyright 2020-2021 OnFinality Limited authors & contributors
```
Comment (ianhe8x) [inline packages/cli/src/template/interface.ts.ejs:1]: we can remove license header for code template. We may own the copyright of template, but not the code generated from the template.

File: packages/cli/src/template/model.ts.ejs [hunk 6]
```diff
@@ -1,6 +1,11 @@
 // Auto-generated , DO NOT EDIT
 import {Entity} from "@subql/types";
 import assert from 'assert';
+<%if (props.importJsonInterfaces.length !== 0) { %>
+import {<% props.importJsonInterfaces.forEach(function(interface){ %>
+    <%= interface %>
```
Comment (ianhe8x) [inline packages/cli/src/template/model.ts.ejs:6]: need comma to split interfaces

File: packages/cli/src/controller/codegen-controller.ts [hunk 7]
```diff
@@ -28,55 +41,128 @@ export interface ProcessedField {
...
+      await renderTemplate(INTERFACE_TEMPLATE_PATH, path.join(typesDir, `interfaces.ts`), interfaceTemplate);
+      exportTypes.interfaces = true;
+    } catch (e) {
+      throw new Error(`When render json interfaces having problems.`);
+    }
+  }
 }
 
 export function processFields(
+  type: string,
```
Comment (ianhe8x) [inline packages/cli/src/controller/codegen-controller.ts:79]: make it an enum if the only possible values are 'entity' and 'json'

File: packages/cli/src/controller/codegen-controller.ts [hunk 8]
```diff
@@ -28,55 +41,128 @@ export interface ProcessedField {
...
+    const injectField = {
       name: field.name,
-      type: newType,
       required: !field.nullable,
-      indexed,
-      unique,
-    });
+      isArray: field.isArray,
+    } as ProcessedField;
+    if (type === 'entity' && indexFields) {
```
Comment (ianhe8x) [inline packages/cli/src/controller/codegen-controller.ts:91]:
```suggestion
    if (type === 'entity') {
```

**Review by ianhe8x — CHANGES_REQUESTED** [review]

**Review by ianhe8x — APPROVED** [review]

**Pull request merged**